kvfeed: send scan requests via RANGEFEED class

pav-kv opened the pull request:
This PR makes changefeed initial scan traffic go through the `RANGEFEED` connection class, by setting the new `rpc.ConnectionClass` parameter in the `BatchRequest` header. Since this extends the API, extra care is taken to treat unknown enum values as `DEFAULT`, e.g. this can be needed in mixed-version clusters.

This PR also introduces a `COCKROACH_RPC_USE_DEFAULT_CONNECTION_CLASS` env variable which overrides all non-`SYSTEM` traffic go through the `DEFAULT` class. This is an escape hatch, in case we need to revert to the previous behaviour.

---

Part of #108992

Release note (performance improvement): this change makes changefeed initial scan traffic, which can be bulky, go through a different RPC/TCP connection than the foreground SQL/KV traffic. This reduces interference between workloads, and reduces chances of head-of-line blocking issues.

Release note (ops change): the new environment variable, `COCKROACH_RPC_USE_DEFAULT_CONNECTION_CLASS`, instructs CRDB to send most network/RPC workloads, except the system traffic, through a single "default" RPC/TCP connection. This is a safety mechanism, it reverts to pre-`v24.1` behaviour in case the environment does not tolerate multiple TCP connections. The default behaviour in `v24.1`+ is having multiple connections dedicated to different types of traffic, such as raft and rangefeeds.

pav-kv committed:
kvfeed: send scan requests via RANGEFEED class

This commit adds an rpc.ConnectionClass parameter to the BatchRequest
header, and sets it to RANGEFEED for changefeed initial scan requests.

Epic: none

Release note (performance improvement): this change makes changefeed
initial scan traffic, which can be bulky, go through a different RPC/TCP
connection than the foreground SQL/KV traffic. This reduces interference
between workloads, and reduces chances of head-of-line blocking issues.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/117810)
<!-- Reviewable:end -->

pav-kv commented on pkg/kv/batch.go:
```diff
@@ -53,6 +53,9 @@ type Batch struct {
 	reqs            []kvpb.RequestUnion
+	// IsBulkFeed specifies whether this batch is a bulk scan request initiated by
+	// a changefeed initial scan.
+	IsBulkFeed bool
```
We could replace this with a check that `len(reqs) == 1` and `reqs[0]` is a `ScanRequest`.

Then we would add a bool or some more generic hint to the `ScanRequest` proto.

erikgrinaker replied:
As discussed on Slack, we'll probably want to add a `ConnectionClass` field in the batch request header, in case the request needs to propagate across the network (consider e.g. follower routing which will proxy via a different node's DistSender).

pav-kv replied:
Done, PTAL.

pav-kv commented on pkg/kv/kvpb/api.proto:
```diff
@@ -71,6 +71,29 @@ enum RoutingPolicy {
+  // scan requests. The rangefeed traffic is potentially bulky, so it has its
+  // dedicated connection class. Changefeed is a major consumer of rangefeed,
+  // and has similar properties, so we place them together.
+  RANGEFEED = 2;
```
NB: value == 2 to match `rpc.RangefeedClass`. This is not required/used, but just in case.

miretskiy commented on pkg/kv/kvpb/api.proto:
```diff
@@ -71,6 +71,29 @@ enum RoutingPolicy {
+  //
+  // As an example, at the time of writing this is used by changefeed initial
+  // scan requests. The rangefeed traffic is potentially bulky, so it has its
+  // dedicated connection class. Changefeed is a major consumer of rangefeed,
```
I don't know if we should over-emphasize changefeed here -- physical cluster replication would also be a big consumer for this (and btw, you shoud probably change that too).

miretskiy commented on pkg/kv/kvpb/api.proto:
```diff
@@ -71,6 +71,29 @@ enum RoutingPolicy {
+  // scan requests. The rangefeed traffic is potentially bulky, so it has its
+  // dedicated connection class. Changefeed is a major consumer of rangefeed,
+  // and has similar properties, so we place them together.
+  RANGEFEED = 2;
```
Should we perhaps call this something else? Like `BULK_DATA` or similar? It's not specific to rangefeed -- we now use that for scan requests for example.

pav-kv replied:
I like the idea. It would be a bit ugly though because this translates to `rpc.RangefeedClass`. Maybe we should generalize/deprecate `rpc.RangefeedClass` first.

@erikgrinaker WDYT?

pav-kv replied:
Added a commit that renames the `rpc` class too.

erikgrinaker replied:
This seems misleading to me, since we don't actually use it for bulk traffic -- imports, schema changes, restores etc. don't use it. Adding a separate bulk class might make sense if we're actually going to use it, but I don't think we should spend time on it now, and I'm not sure it makes sense to send rangefeeds across the bulk class even if we have one (some customers may not want bulk traffic to increase feed latency, nor do we for our internal system rangefeeds).

I'd be inclined to just keep `RangefeedClass` for these, since that's the only thing it will be used for in 24.1 (fine, initial scans aren't rangefeeds, but initial scans are only used together with rangefeeds so that seems fine to me).

pav-kv replied:
Reverted to using the rangefeed class. We'll generalize later if needed.

miretskiy reviewed (commented):
I think the PR description needs to be explanded significantly, with release notes.
We should clearly explain why this is done (splitting bulky traffic over dedicated http2 connection to reduce head-of-line blocking for other types of traffic).

pav-kv commented:
> I think the PR description needs to be explanded significantly, with release notes.

Yep, I'll write the description when we flesh out the details.

pav-kv commented:
@erikgrinaker Here is the `diff` between `BatchRequest` memory layouts ([obtained](https://github.com/dominikh/go-tools/tree/master/cmd/structlayout) with this tool) before/after:

```
diff before.txt after.txt                                                                                                   initial-scan-via-rangefeed-class!?
42c42,43
< padding: 217-224 (size 7, align 0)
---
> padding: 217-220 (size 3, align 0)
> BatchRequest.Header.ConnectionClass github.com/cockroachdb/cockroach/pkg/kv/kvpb.ConnectionClass: 220-224 (size 4, align 4)
```

So, adding this field is "free" for the in-memory overhead.

pav-kv commented on pkg/ccl/changefeedccl/kvfeed/scanner.go:
```diff
@@ -191,6 +191,7 @@ func (p *scanRequestScanner) exportSpan(
 		r := kvpb.NewScan(remaining.Key, remaining.EndKey).(*kvpb.ScanRequest)
 		r.ScanFormat = kvpb.BATCH_RESPONSE
 		b.Header.TargetBytes = targetBytesPerScan
+		b.Header.ConnectionClass = kvpb.ConnectionClass_BULK_DATA
```
Should we hook this in to the [env var](https://github.com/cockroachdb/cockroach/blob/9b10f083f8889cdc1bc549d9428f06fd784ba081/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L58-L66) that overrides to the default class?

We could also add a separate env var for this here.

More generally, we already have 2 env vars for overriding connection classes. Since this is a third candidate for such a setting, I wonder if we should generalize all these cases into a single env var that's a key-value list of all the overrides. Something like:

`COCKROACH_USE_DEFAULT_CONNECTION_CLASS=raft,rangefeed` - resets to default class for rangefeed and raft traffic.
`COCKROACH_USE_DEFAULT_CONNECTION_CLASS=*` - resets to default class for all traffic types in question.

pav-kv replied:
Wrote a quick and dirty prototype: #117882.

erikgrinaker replied:
I don't think we need to overdo it, we mostly need this as an escape hatch in case something breaks with the separate connection classes (not sure what that would be -- maybe some environments have limits on concurrent TCP connections).

I think a common `COCKROACH_USE_DEFAULT_CONNECTION_CLASS=true|false` would go a long way.

pav-kv replied:
Added another commit with this env var that reset every. It completely resets the behaviour to the previous state when we only had "default" and "system".

pav-kv commented:
@erikgrinaker This is ready for another pass.

erikgrinaker commented on pkg/kv/kvpb/api.proto:
```diff
@@ -71,6 +71,27 @@ enum RoutingPolicy {
+// differ. This enum only defines classes used for BatchRequest traffic, and
+// only defines preferences. There are more classes in rpc package that are not
+// relevant here, such as RaftClass.
+enum ConnectionClass {
```
Can we replace `rpc.ConnectionClass` with this enum? It seems unfortunate to have two separate enums that mean the same thing and need to be kept in sync.

pav-kv replied:
Agree, but they are not quite the same thing. Logically, this enum is neither a subset nor a superset of `rpc.ConnectionClass` (even though at the moment it is). Generally, it is a "tag".

Consider a future in which we may categorize more types of streams/traffic (schema changes, imports/exports, etc). We don't necessarily want to have that many RPC classes - some of them can be multiplexed to "bulk" class or alike. In some more distant future with QUIC, these tags can translate to separate streams more fearlessly.

Such tags are also useful for other stuff, like observability/monitoring. If we have 10 tags, we can break down the "bytes/s" graphs into 10 components. Even though all 10 can be flowing through the same class / TCP connection.

This `enum` is part of the `BatchRequest` API (which goes into the wire too), so is limited to connection classes that are used for `Send`s (and e.g. does not include `RaftClass`). It basically virtualizes away the lower-level RPC stuff - it tells the "kind" of traffic, and RPC layer chooses the physical implementation for it.

I'm not super inclined to reverse the dependency here and say that RPC depends on protos in `kvpb`. But maybe putting this into some `rpcpb` would be fine for now.

erikgrinaker replied:
We can build something like that if and when we need it. Let's just keep this simple and have a single concept of connection classes instead of adding yet another layer of indirection (aka complexity) that we may or may not need at some unknown point in the future.

>I'm not super inclined to reverse the dependency here and say that RPC depends on protos in kvpb.

Why not? All sorts of things depend on `kvpb` or `roachpb`. The only real difference is that they can be used in Protobufs.

Either way, let's just get this wrapped up. I'll trust your judgement on how to structure the enums.

pav-kv replied:
Done. Moved `ConnectionClass` to a new `rpcpb` package, and left aliases for the old type and constants, so that we don't need to touch many tens of usages in this PR. Added a TODO to get rid of the aliases later.

erikgrinaker reviewed (commented):
Approach seems good, modulo class bikeshedding.

pav-kv committed:
rpc: introduce rpcpb package with ConnectionClass

This commit adds a pkg/rpc/rpcpb package, and puts the ConnectionClass
enum in it. This enum needs to be in a proto because we are going to add
it to kvpb.BatchRequest API.

This adds a new Go package rather than extends the existing pkg/rpc to
minimize dependencies that importing this proto will bring with it. At
the moment, pkg/rpc already indirectly imports kvpb, so importing from
kvpb would require dealing with circular dependencies first.

Epic: none
Release note: none

pav-kv commented on pkg/kv/kvpb/api.proto:
```diff
@@ -2883,9 +2884,20 @@ message Header {
+  // the SYSTEM class.
+  //
+  // If the enum value is unknown (for instance, may happen when the cluster is
+  // in a mixed-version state after a new value got added), DEFAULT is assumed.
```
The `DEFAULT` safety is not implemented yet, the values are passed as is.

Maybe we don't need to reset to `DEFAULT`, just assume that all values of this are acceptable. It's not safe though, to things that depend on [NumConnectionClasses](https://github.com/cockroachdb/cockroach/blob/8484d4bf6af7866955183fa7223b30084f511697/pkg/rpc/connection_class.go#L45) being greater than the max connection class.

pav-kv replied:
I'd like to avoid having migrations boilerplate later in case we want to add a new type.

pav-kv replied:
The downside of option (1) is that the initiator of the request has very little control of which class will end up being used on the retransmitting nodes.

Example: suppose we want to split off some portion of `RAFT` class into a subset (e.g. `RAFT_SNAPSHOTS`). We don't want nodes that don't know `RAFT_SNAPSHOT` yet to suddenly switch to the `DEFAULT` class. A good behaviour would be to default to the "old" `RAFT` in this case.

Option (3) gives best control of this, because the initiator of the request can set `RAFT` or `RAFT_SNAPSHOT` depending on the version gate state.

Option (2) is somewhat worse, but at least we know that the request will go through a separate connection.

Option (1) is worst, because it can break old assumptions about what goes where.

miretskiy replied:
Isn't this connection class business purely a client side specific construct and does not effect
the server side in any way and thus likely doesn't not have any mix-version case issues?
The client either knows that there is a new "RAFT_SNAPSHOT" class, or it doesn't -- and it's
the client that initiates the connections.

 I understand that (unfortunately) this information is part of kvpb header -- and so theoretically
this connection class could be used by the server side which would raise mix-version issues.
(I say unfortunately because, well.. it's unfortunate because this information is connection
specific and not KV request specific.  It is howerver, very convienient place and at least
somewhat similar in spirit to "admission" header stuff, so -- that's fine (imo)).

Is the above a concern (meaing the use of this connection class on the server side)?  
Could this header field be marked (in the comment) as "client side only"and just be done with it?
Can we punt this until we actually have a situation where the server cares about
connection class (perhaps because it tries to establish its own rpc connections?)

Ragardless, connection class, I feel is an optimization and is a best effort.
If we want the server to support "unknown" connection classes, then we need to have the header
use an "int" instead of strongly typed enum to represent connection class; and the code
that translates it to well defined connection classes, with a fallback to default  doesn't sound that bad to me.

pav-kv replied:
My understanding is that `Send(BatchRequest)` can be transparently retransmitted by other nodes. So there is no exactly one client-server here.

Quoting Erik:
> Consider e.g. the case of follower routing, where we'll proxy via some other node's `DistSender` -- we'll want it to use the same connection class.

pav-kv replied:
> If we want the server to support "unknown" connection classes, then we need to have the header
use an "int" instead of strongly typed enum to represent connection class; and the code
that translates it to well defined connection classes, with a fallback to default doesn't sound that bad to me.

I agree, also leaning towards not taking option (2). Using option (1) with assuming that it's possible we will need option (3) some day seems like a good compromise.

pav-kv replied:
Added a commit implementing this.

erikgrinaker reviewed (approved):
LGTM, thanks!

pav-kv committed:
rpc: convert unknown connection classes to DEFAULT

Since the ConnectionClass can now be received over RPCs, it is possible
that the receiver node does not know/support it, e.g. in a mixed-version
cluster.

This commit is a defence-in-depth against such cases: fall back to the
DEFAULT class for the unknown inputs. This is also a convenient
behaviour if we decide to split any more traffic from the DEFAULT class.

For better control over this behaviour, use version gates.

Epic: none
Release note: none

pav-kv committed:
rpc: add env var for resetting to default conn class

This commit introduces a COCKROACH_RPC_USE_DEFAULT_CONNECTION_CLASS env
variable, which instructs CRDB to use the DEFAULT connection class for
all traffic that otherwise would use a separate connection class, such
as RAFT.

This option reverts the behaviour to pre-v24.1 state, and is an escape
hatch in case something breaks with the separate connection classes
(could be anything, e.g. an environment has limits on concurrent TCP
connections).

Epic: none

Release note (ops change): the new environment variable,
COCKROACH_RPC_USE_DEFAULT_CONNECTION_CLASS, instructs CRDB to send most
network/RPC workloads, except the system traffic, through a single
"default" RPC/TCP connection. This is a safety mechanism, it reverts to
pre-v24.1 behaviour in case the environment does not tolerate multiple
TCP connections. The default behaviour in v24.1+ is having multiple
connections dedicated to different types of traffic, such as raft and
rangefeeds.

pav-kv commented:
TYFRs @erikgrinaker @miretskiy

I added release notes and description. I will merge this on Monday, in case you want to take a final look (I added a couple of commits since the last reviews).

miretskiy reviewed (approved):
Thanks for updated PR commits; <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

erikgrinaker commented on pkg/rpc/connection_class.go:
```diff
@@ -76,3 +76,19 @@ func ConnectionClassForKey(key roachpb.RKey, def ConnectionClass) ConnectionClas
+//
+// If the simple behaviour of falling back to DefaultClass is not acceptable, a
+// version gate must be used to transition between versions correctly.
+func ConnectionClassOrDefault(c ConnectionClass) ConnectionClass {
```
Bummer, I thought gogoproto would coerce unknown values to the zero value by default. Did you check this?

pav-kv replied:
Proto [spec](https://protobuf.dev/programming-guides/proto3/#enum) says that unknown values are left as is:

> During deserialization, unrecognized enum values will be preserved in the message, though how this is represented when the message is deserialized is language-dependent. In languages that support open enum types with values outside the range of specified symbols, such as C++ and Go, the unknown enum value is simply stored as its underlying integer representation. In languages with closed enum types such as Java, a case in the enum is used to represent an unrecognized value, and the underlying integer can be accessed with special accessors. In either case, if the message is serialized the unrecognized value will still be serialized with the message.

erikgrinaker commented on pkg/kv/kvclient/kvcoord/dist_sender.go:
```diff
@@ -2287,13 +2287,11 @@ func (ds *DistSender) sendToReplicas(
-	// states gracefully.
-	rpcClass := rpc.ConnectionClassOrDefault(ba.ConnectionClass)
-	// Upgrade the connection class to SYSTEM, for critical ranges.
-	rpcClass = rpc.ConnectionClassForKey(desc.RSpan().Key, rpcClass)
```
nit: should this change be in the previous commit?

pav-kv replied:
Done.

pav-kv commented:
bors r=erikgrinaker,miretskiy
